nixos/hercules-ci-agent: init

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Make Hercules CI Agent easier to install and configure.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

roberth committed:
nixos/hercules-ci-agent: init

flokli commented on nixos/modules/misc/ids.nix:
```diff
@@ -647,6 +648,7 @@ in
       paperless = 315;
       #mailman = 316;  # removed 2019-08-30
       zigbee2mqtt = 317;
+      hercules-ci-agent = 318;
```
Do we need to allocate a uid and gid, or can we make use of `DynamicUser=yes` here?

roberth replied:
I don't think static allocation is strictly necessary. `DynamicUser` doesn't seem to be an option because it is not compatible with setuid workers. Although those aren't implemented yet, going back and forth will require the migration of state directories, twice.

flokli replied:
setuid workers seem to be a valid usecase for not using `DynamicUser` - but there's no need for static uid allocation.
Please only create the user using `users.users.hercuies-ci-agent` (without at static uid), and only if it's not the default.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,211 @@
+
+  format = pkgs.formats.toml {};
+
+  checkNix =
```
I'm not a fan of something failing at build time here.

What about checking the nix version and failing if it's too low at evaluation time, and prompting users to use `nixUnstable` in that case?

roberth replied:
I would like to avoid depending on unreleased versions, which is why I've created the option to apply a minimal patch on the released version.
Deriving this information from options only seems a bit risky. For example if the user has their own patched Nix, how do we check that it's ok?

It's only a temporary solution. Once 2.4 is released, I'm happy to reduce this to a simple eval-only check.

flokli replied:
Hmmh, this code also depends on unreleased versions. Contrary to `nixUnstable`, you build your own version on-the-fly here, which is even less risky for sudden build failure.

I'd assume Nix 2.4 will be released before 19.10 - mind just recommending `nixUnstable` in an evaluation failure/warning (depending on how required it is) here if `!(lib.versionAtLeast (lib.getVersion config.nix.package))`?

roberth replied:
The patch has been tested since 2.3.4 and has not needed any changes for the total of four releases. It's a patch on the stable branch after all. 
I'd be happy to remove the patching when 2.4 is released. But I wouldn't just assume it's released before NixOS.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,211 @@
+        If you set this to false, please keep up with the change log.
+      '';
+    };
+    baseDirectory = mkOption {
+      type = types.path;
+      default = "/var/lib/hercules-ci-agent";
+      description = "State directory (secrets, work directory, etc) for agent";
+    };
```
Can you use `StateDirectory=` in the unit?

roberth replied:
This maps to an entry in the config file. I suppose I could derive `StateDirectory` from it, but I'm not sure what would be the benefit of doing so.

flokli replied:
Hmm, I wasn't specific here - if this were using `DynamicUser` (which it can't, due to suid support), We could just have defaulted the state directory to this location.

It's probably still beneficial to distinguish between the home directory of the user running the service (which could be somewhere else than `/var/lib/buildkite-agent`), and state files (which should always live in `/var/lib/$serviceName`.

roberth replied:
It currently depends on dotfiles to configure automatic S3 push for example, so the default home `/var/empty` wouldn't work. Disallowed or transient dotfiles are normally the benefit I suppose. Is there another benefit that could offset the migration that will be required for current installations?

roberth replied:
Relevant issue: https://github.com/hercules-ci/hercules-ci-agent/issues/239
For now I don't see a reason to use `StateDirectory`.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,211 @@
+      default = "/var/lib/hercules-ci-agent";
+      description = "State directory (secrets, work directory, etc) for agent";
+    };
+    user = mkOption {
+      description = "Unix system user that runs the agent service";
+      type = types.str;
+    };
```
Does this need to be configurable, or can users just override the unit?

roberth replied:
This means I can't automatically turn of the user creation, but that seems to be the norm. Trying to do it results in infinite recursions.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/default.nix:
```diff
@@ -0,0 +1,99 @@
+      };
+    };
+
+    systemd.services.hercules-ci-agent-restarter = {
+      serviceConfig.Type = "oneshot";
+      script = ''
+        # Wait a bit, with the effect of bundling up file changes into a single
+        # run of this script and hopefully a single restart.
+        sleep 10
+        if systemctl is-active --quiet hercules-ci-agent.service; then
+          if ${testCommand}; then
+            systemctl restart hercules-ci-agent.service
+          else
+            echo 1>&2 "WARNING: Not restarting agent because config is not valid at this time."
+          fi
+        else
+          echo 1>&2 "Not restarting hercules-ci-agent despite config file update, because it is not already active."
+        fi
+      '';
+    };
```
Can you explain a bit why this is necessary?

roberth replied:
Changes in the secrets do not affect the unit in any way that would case a restart, which is currently necessary to reload the secrets.

flokli replied:
Please add a comment to above this, so people don't need to dig up this PR, but see if next to the code.

Secret handling, and reload behaviour are somewhat intertwined with https://github.com/NixOS/rfcs/pull/59 - I'd feel much better if hercules-ci wouldn't implement its own thing here, but behaved similar to other NixOS services - which currently don't magiclaly reload on file changes (at least if the daemon itself doesn't watch these files and does so on its own).

roberth replied:
The reload behavior was expected by users, as shown by support requests. The nix-darwin module has the same reloading, which is trivially easy with launchd.
I'm planning to integrate reloading into the agent executable but for now the gap can be bridged by this unit.

I'll add the comment.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/default.nix:
```diff
@@ -0,0 +1,99 @@
+    nix.trustedUsers = [ cfg.user ];
+    services.hercules-ci-agent.extraOptions.nixUserIsTrusted = true;
+
+    users.users.hercules-ci-agent = mkIf (cfg.user == defaultUser) {
+      home = cfg.baseDirectory;
+      createHome = true;
+      group = "hercules-ci-agent";
+      uid = config.ids.uids.hercules-ci-agent;
+      description = "Hercules CI Agent system user";
+      isSystemUser = true;
+    };
+
+    users.groups.hercules-ci-agent = mkIf (cfg.user == defaultUser) {
+      gid = config.ids.gids.hercules-ci-agent;
+    };
```
This is probably not needed if we use `DynamicUser`.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/deploy-keys.nix:
```diff
@@ -0,0 +1,73 @@
+{ config, lib, options, pkgs, ... }:
```
We shouldn't read nixops config options inside nixpkgs IMHO. Can we just ask for a location to certain files containing secrets (with sane defaults), and ask the user however they seem fit?

roberth replied:
I'm not sure if I understand correctly, because it's not really reading as much as writing nixops configuration.
Are you suggesting to provide options such that

    services.hercules-ci-agent.nixops.clusterJoinToken.keyFile = ./cluster-join-token;
    services.hercules-ci-agent.nixops.binaryCachesJson.keyFile = ./binary-caches.json;

cause the appropriate NixOps configs to be set?

Or do you disapprove of the entire concept of setting NixOps options for the user? I do think this is a rather nice possibility with NixOS/NixOps to make things a bit easier to set up.

roberth replied:
I've made it opt-in and the options now serve the dual purpose of documenting the default location for these files.

flokli replied:
Reading nixops options from inside nixpkgs is really a mispattern - how can we ever refactor things without breaking it on the other side?

NixOS modules really should just provide options to configure services to file locations (with sane defaults), and don't contain any deployment-tool specific code.

https://github.com/NixOS/rfcs/pull/59 proposes a generic API for these things, but until there's progress on that front, I'd really suggest providing options like `services.hercules-ci-agent.clusterJoinTokenFile` and `services.hercules-ci-agent.binaryCachesJsonFile` - see the gitlab module for example.

roberth replied:
If your change the interface of nixops, doesn't that break all consumers and isn't that easier to fix in NixOS than countless user deployments?
To answer the question literally: you can refactor by adding compatibility logic that reads the `options` module parameter, where all options can be introspected, including those from nixops.

`services.hercules-ci-agent.clusterJoinTokenFile` etc is already available via `extraOptions` if you need something custom for some reason.

I find this "mispattern" position rather disappointing. Hercules CI users are expecting this high-level behavior from the out-of-tree module now. Do I really have to bother them to needlessly think about those details again? It seems like a step backwards.

https://github.com/NixOS/rfcs/pull/59 doesn't seem like a sufficiently mature solution just yet. I *might* be able to help the RFC a bit, but I don't think it should be a blocker.

flokli replied:
> I find this "mispattern" position rather disappointing. Hercules CI users are expecting this high-level behavior from the out-of-tree module now. Do I really have to bother them to needlessly think about those details again? It seems like a step backwards.

There's currently no in-tree module using `deployment.keys`, and "it worked that way in an out-of-tree module" isn't really a justification by itself IMHO. Often things need to be refactored when sent mainline, to be more consistent with how things are done there.

Having the module reading secrets from a specific path by default (which is documented in the description and release notes/migration guide) seems less of a problem than using an option that isn't even documented in `man configuration.nix` and fails `nixos-rebuild`, when setting it on the machine:

```
error: The option `deployment' defined in `/etc/nixos/configuration.nix' does not exist.
(use '--show-trace' to show detailed location information)
```

Also, relying on things provided by nixops, or reading stuff from inside nixops has proven to be super painful. (See https://github.com/NixOS/nixops/pull/1156 and linked issues for details).

I assume people currently using the out-of-tree module need to do some sort of migration anyways (at least stop importing the out-of-tree module), so doing some more cleanup on the way shouldn't hurt.

roberth replied:
>  fails nixos-rebuild, when setting it on the machine:
> ```
> error: The option `deployment' defined in `/etc/nixos/configuration.nix' does not exist.
> (use '--show-trace' to show detailed location information)
> ```

fyi that was taken care of.

> Also, relying on things provided by nixops, or reading stuff from inside nixops has proven to be super painful.

At least I could fix it in a single place for everyone. It's even possible to adapt `config` to the `options` provided as was shown. Nonetheless I've decided not to continue the pioneering, because even if successful the cost/benefit seems a bit off.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/gc.nix:
```diff
@@ -0,0 +1,24 @@
+
+in
+{
+  options.services.hercules-ci-agent = {
```
This is not specific to hercules-ci-agent at all - it only enables some `nix.gc` options. If you you feel like more options are needed there, please add them directly to the `nix.gc` namespace.

roberth replied:
It is required for correct operation though. That's not to say that it can't be done differently, which is why it's opt-in. That's already a change from the existing module.
Another reason for this to exist is that the agent will implement garbage collection logic to purge paths which are unlikely to be used first.

I'll add a comment.

flokli replied:
If it's required for correct operation, why not assert some of the `nix.gc` options being set, and exposing the `--max-freed` option in the `nix.gc` namespace (and module) as an option?

You shouldn't cross boundaries of this module, and it's something useful for all NixOS users, not just people configuring the `hercules-ci-agent`.

roberth replied:
It's potentially useful for all NixOS users, but I'm not convinced that this should be the blessed implementation.
I've removed it.

flokli commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,207 @@
+    then lib.warn "Hercules CI Agent module will not patch 2.4 pre-release. Make sure it includes (equivalently) PR #3043, commit d048577909 or is no older than 2020-03-13." pkgs.nix
+    else pkgs.nix.overrideAttrs (
+      o: {
+        patches = (o.patches or []) ++ [ ./issue-3398-path-info-cache-ttls-backport-2.3.patch ];
```
[If this stays](https://github.com/NixOS/nixpkgs/pull/94804/files#r466724842), it should be `fetchpatch`ed, instead of shipped inside nixpkgs.

aanderse reviewed (changes requested):
This seems to break a number of conventions, deviating from accepted NixOS module standards. @flokli mentioned a number of these and I wonder if this module is more appropriate for NUR instead. If there is motivation to get this upstream (which I think is a good idea, this module provides high value) I think the comments @flokli left are important.

roberth commented:
I've made the changes to turn this into a conventional module.
This now also uses RFC 42-style `settings` with freeform modules (thanks @Infinisil !)

flokli commented:
Thanks, this now looks way more appropriate :-)

I still think the "patching nix if required" part doesn't belong into the module - raise an erorr if the version of Nix isn't recent enough (or grep the header files if the version string isn't enough).

Infinisil commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,215 @@
+  imports = [
+    (mkRenamedOptionModule ["services" "hercules-ci-agent" "extraOptions"] ["services" "hercules-ci-agent" "settings"])
+    (mkRenamedOptionModule ["services" "hercules-ci-agent" "baseDirectory"] ["services" "hercules-ci-agent" "settings" "baseDirectory"])
+    (mkRenamedOptionModule ["services" "hercules-ci-agent" "concurrentTasks"] ["services" "hercules-ci-agent" "settings" "concurrentTasks"])
```
Why these entries for a new module? Same for the option rename in `default.nix`

roberth replied:
They're for those migrating from a pre-existing out-of-tree module.

Infinisil commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,215 @@
+      description = ''
+        Enable to run Hercules CI Agent as a system service.
+
+        <link xlink:href="https://hercules-ci.com">Hercules CI</link> is a
```
I tend to prefer the `<link xlink:href="https://hercules-ci.com"\>` links because these will also be visible in the man page.

roberth replied:
Good point. In this specific case I think making it a URL just adds noise, but further down I've done it to a 'deep' link. Thanks!

Infinisil commented on nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix:
```diff
@@ -0,0 +1,215 @@
+      # even shortly after the previous lookup. This *also* applies to the daemon.
+      narinfo-cache-negative-ttl = 0
+    '';
+    nix.package = mkIf cfg.patchNix patchedNix;
```
It might be better to use an overlay for this instead, because it will allow users to still specify e.g. `nix.package = pkgs.nixFlakes`, and the overlay can patch all of `pkgs.{nix,nixUnstable,nixFlakes}` (if necessary for each of them).

roberth replied:
It's only necessary for the daemon and only when its `2.3`. They can set it to `nixFlakes` just fine because that version does not need patching.
It's a temporary situation until `2.4` is released. At that point, I'll drop the whole patching business, so there's little point in improving this.

Infinisil replied:
Ah so the intended flow is:
- User enables module
- Because `checkNix = true` by default and `patchNix = false`, user gets error that the Nix check failed and they need to use a version with a patch
- User sets `patchNix = true`, therefore overriding the default Nix version (or I guess it would error if the user also assigned `nix.package` in their config)

I guess that's fine

Infinisil commented on nixos/modules/services/continuous-integration/hercules-ci-agent/default.nix:
```diff
@@ -0,0 +1,85 @@
+          if ${testCommand}; then
+            systemctl restart hercules-ci-agent.service
+          else
+            echo 1>&2 "WARNING: Not restarting agent because config is not valid at this time."
```
How about using `testCommand` in the main services `preStart` script, failing if it's invalid? I believe this would make system activation fail already

roberth replied:
Done.

infinisil reviewed (changes requested):
Lovely to see some nice use of the recent module additions!

roberth commented:
@flokli @aanderse @Infinisil 

I have addressed all comments.

The patching options are useful until 2.4 is released, so I suggest they stick around but only until then.
Any final comments?

infinisil reviewed (approved):
Looks good to me! The checking and patching is indeed a bit gnarly, but I think that's fine for a temporary convenience improvement
